### PR TITLE
Update model.py

### PR DESCRIPTION
**Pull request by tianfengyijiu**
fix bugs

### CODE REVIEW

**Comment by wizyoung** [comment]
Thanks again. I will review these changes when I'm available (maybe this weekend)

**Comment by factplay1** [comment]
@wizyoung Can you pls review and merge the bug in the loss function? When I train the model from scratch, I can't achieve the same accuracy you have reported.

**Comment by wizyoung** [comment]
I'll not approve this PR. Because it uses tf.py_func to do for-loop inside and the gradients will not be propagated properly. The correct way is to use tf.while_loop instead.

**Comment by factplay1** [comment]
Can you try to update that your self then?

**Comment by wizyoung** [comment]
@factplay1 I wish I had time. But god knows how busy and tired I'm during these days. Sorry about that.

**Comment by factplay1** [comment]
Your work has been very inspirational and useful for lots of people like me. I would very much appreciate if you could give it a try over the weekend perhaps and see if you can push a fix? :)

**Comment by wizyoung** [comment]
@factplay1 I plan to push a fix today :)

**Comment by factplay1** [comment]
@wizyoung   Thanks so much, really appreciate your time :)
I can confirm that training has a much better progress than before. I've started two jobs on training CoCo from scratch, and from finetuning using the darknet Yolo-body, they are learning better than before.

By the way, can you pls tell me what's the mAP @50 you have got on CoCo? Is that from scratch training, or loading the darknet weights?

Thanks :)

**Comment by wizyoung** [comment]
@factplay1 I have not tested on the COCO dataset. I made two quick tests on VOC dataset. (1) loading darknet weights I got ~77 mAP with 416*416 input; (2) loading weighs from all layers except last three conv2d layers in 3 scales, I got ~84 mAP with 544*544 input. No try-hard fine-tuning. Performance seems normal.

Now it seems there is no major bug in my code. But there are still some improvements and I plan to update, like:
(1) resize image changes to letterbox resize (finished but not pushed yet)
(2) loading weights from all layers except last three conv2d layers in 3 scales (finished but not pushed yet)
(3) try adding exponential moving average decay for weights (still in plan)

**Comment by wizyoung** [comment]
@factplay1 Really appreciate your participation. But currently I have no plan to train on COCO dataset.

**Comment by factplay1** [comment]
@wizyoung   Thanks a lot for your input. Very much appreciate your great work.
My experiments with CoCo are progressing normally I think. I'll report if I came across something abnormal.

Looking forward to your new update on "exponentially moving average weights".

Cheers,

**Comment by wizyoung** [comment]
This weekend I read a lot of C source codes from the original darknet repo, and it seems the current model design is consistent with the original repo. So I'll close this PR. Today I'll release a new version to fix other small bugs.

In the future, I plan to add the YOLOV3-SPP, YOLOV3-tiny versions or introduce other backbones if I have free time.
